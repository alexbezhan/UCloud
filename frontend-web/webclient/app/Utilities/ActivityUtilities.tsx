import {ActivityFilter} from "Activity";
import {ScrollRequest} from "Scroll";
import {buildQueryString} from "./URIUtilities";

interface ParamsType extends Omit<ActivityFilter, "maxTimestamp" | "minTimestamp">, Partial<ScrollRequest<number>> {
    maxTimestamp?: number;
    minTimestamp?: number;
}

export const activityQuery = (scroll: ScrollRequest<number>, filter?: ActivityFilter) => {
    const params: ParamsType = {};
    if (!!scroll.offset) params.offset = scroll.offset;
    params.scrollSize = scroll.scrollSize;

    if (filter !== undefined) {
        if (filter.type !== undefined) params.type = filter.type;
        if (filter.collapseAt !== undefined) params.collapseAt = filter.collapseAt;
        if (filter.maxTimestamp !== undefined) params.maxTimestamp = filter.maxTimestamp.getTime();
        if (filter.minTimestamp !== undefined) params.minTimestamp = filter.minTimestamp.getTime();
        if (filter.user !== undefined) params.user = filter.user;
    }

    return buildQueryString("/activity/browse", params);
};

export const activityStreamByPath = (path: string) =>
    `/activity/by-path?path=${encodeURIComponent(path)}`;
