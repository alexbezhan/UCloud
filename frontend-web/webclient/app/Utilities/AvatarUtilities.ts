export const saveAvatarQuery = "/avatar/update";
export const findAvatarQuery = "/avatar/find";
