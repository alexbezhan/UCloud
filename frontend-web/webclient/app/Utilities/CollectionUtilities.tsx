export function removeEntry<T>(arr: T[], index: number): T[] {
    return arr.slice(0, index).concat(arr.slice(index + 1));
}

export function setsDiffer<T>(s1: Set<T>, s2: Set<T>): boolean {
    if (s1.size !== s2.size) return true;
    let differ = false;
    s1.forEach(it => {
        if (!s2.has(it)) differ = true;
    });
    return differ;
}

type PrimitiveDataTypes = string | number | boolean;

/**
 * @param set
 * @param entry
 */
export function addEntryIfNotPresent(set: Set<PrimitiveDataTypes>, entry: PrimitiveDataTypes): boolean {
    const size = set.size;
    set.add(entry);
    return size !== set.size;
}

export function groupBy<T>(items: T[], keySelector: (t: T) => string): Record<string, T[]> {
    const result: Record<string, T[]> = {};
    items.forEach(item => {
        const key = keySelector(item);
        const newList = result[key] ?? [];
        newList.push(item);
        result[key] = newList;
    });
    return result;
}

export function associateBy<T>(items: T[], keySelector: (t: T) => string): Record<string, T> {
    const result: Record<string, T> = {};
    items.forEach(item => {
        const key = keySelector(item);
        result[key] = item;
    });
    return result;
}

export function takeLast<T>(items: T[], numberOfItems: number): T[] {
    return items.slice(Math.max(0, items.length - numberOfItems));
}
