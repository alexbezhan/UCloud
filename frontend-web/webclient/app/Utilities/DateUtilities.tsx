import {format} from "date-fns/esm";

// Could potentially cause issues with time if user is outside CEST
export const dateToString = (date: number) => format(date, "HH:mm dd/MM/yyyy");
