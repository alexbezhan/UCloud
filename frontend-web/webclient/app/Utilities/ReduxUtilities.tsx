import activity from "Activity/Redux/ActivityReducer";
import * as AppRedux from "Applications/Redux";
import analyses from "Applications/Redux/AnalysesReducer";
import applications from "Applications/Redux/BrowseReducer";
import detailedApplicationSearch from "Applications/Redux/DetailedApplicationSearchReducer";
import * as TaskRedux from "BackgroundTasks/redux";
import dashboard from "Dashboard/Redux/DashboardReducer";
import {initObject} from "DefaultObjects";
import detailedFileSearch from "Files/Redux/DetailedFileSearchReducer";
import fileInfo from "Files/Redux/FileInfoReducer";
import header, {CONTEXT_SWITCH, USER_LOGIN, USER_LOGOUT} from "Navigation/Redux/HeaderReducer";
import sidebar from "Navigation/Redux/SidebarReducer";
import status from "Navigation/Redux/StatusReducer";
import notifications from "Notifications/Redux/NotificationsReducer";
import * as ProjectRedux from "Project/Redux";
import {Action, AnyAction, combineReducers, createStore, Store} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import {createResponsiveStateReducer, responsiveStoreEnhancer} from "redux-responsive";
import simpleSearch from "Search/Redux/SearchReducer";
import {responsiveBP} from "ui-components/theme";
import uploader from "Uploader/Redux/UploaderReducer";
import avatar from "UserSettings/Redux/AvataaarReducer";
import hookStore from "Utilities/ReduxHooks";

export function configureStore(
    initialObject: ReduxObject,
    reducers,
    enhancers?
): Store<ReduxObject, AnyAction> {
    const combinedReducers = combineReducers<ReduxObject, AnyAction>(reducers);
    const rootReducer = (state: ReduxObject, action: Action): ReduxObject => {
        if ([USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH].some(it => it === action.type)) {
            state = initObject();
        }
        return combinedReducers(state, action);
    };
    return createStore<ReduxObject, AnyAction, {}, {}>(rootReducer, initialObject, composeWithDevTools(enhancers));
}

export const responsive = createResponsiveStateReducer(
    responsiveBP,
    {infinity: "xxl"}
);

export const store = configureStore(initObject(), {
    activity,
    dashboard,
    analyses,
    applications,
    header,
    status,
    sidebar,
    uploader,
    notifications,
    simpleSearch,
    detailedFileSearch,
    detailedApplicationSearch,
    fileInfo,
    hookStore,
    ...AppRedux.reducers,
    avatar,
    loading,
    tasks: TaskRedux.reducer,
    project: ProjectRedux.reducer,
    responsive: createResponsiveStateReducer(
        responsiveBP,
        {infinity: "xxl"}),
}, responsiveStoreEnhancer);

function loading(state = false, action: {type: string}): boolean {
    switch (action.type) {
        case "LOADING_START":
            return true;
        case "LOADING_END":
            return false;
        default:
            return state;
    }
}
