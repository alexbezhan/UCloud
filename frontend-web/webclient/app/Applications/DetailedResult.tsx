import * as React from "react";
import PromiseKeeper from "PromiseKeeper";
import {Cloud} from "Authentication/SDUCloudObject";
import {shortUUID, errorMessageOrDefault} from "UtilityFunctions";
import {Link} from "react-router-dom";
import {connect} from "react-redux";
import {setLoading, updatePageTitle} from "Navigation/Redux/StatusActions";
import {
    DetailedResultProps,
    DetailedResultState,
    StdElement,
    DetailedResultOperations,
    AppState,
    WithAppInvocation
} from ".";
import {fileTablePage} from "Utilities/FileUtilities";
import {hpcJobQuery, cancelJobDialog, inCancelableState, cancelJob} from "Utilities/ApplicationUtilities";
import {Dispatch} from "redux";
import {Dropdown, DropdownContent} from "ui-components/Dropdown";
import {Flex, Box, List, Card, ContainerForText, ExternalLink, Button} from "ui-components";
import {Step, StepGroup} from "ui-components/Step";
import styled from "styled-components";
import {TextSpan} from "ui-components/Text";
import Icon from "ui-components/Icon";
import {setRefreshFunction} from "Navigation/Redux/HeaderActions";
import * as Heading from "ui-components/Heading";
import {JobStateIcon} from "./JobStateIcon";
import {MainContainer} from "MainContainer/MainContainer";
import {SnackType} from "Snackbar/Snackbars";
import {snackbarStore} from "Snackbar/SnackbarStore";
import LoadingIcon from "LoadingIcon/LoadingIcon";
import {Spacer} from "ui-components/Spacer";
import {EmbeddedFileTable} from "Files/FileTable";

const Panel = styled(Box)`
    margin-bottom: 1em;
`;

Panel.displayName = "Panel";

class DetailedResult extends React.Component<DetailedResultProps, DetailedResultState> {
    private stdoutEl: StdElement;
    private stderrEl: StdElement;

    constructor(props: Readonly<DetailedResultProps>) {
        super(props);
        this.state = {
            complete: false,
            appState: AppState.VALIDATED,
            status: "",
            stdout: "",
            stderr: "",
            stdoutLine: 0,
            stderrLine: 0,
            stdoutOldTop: -1,
            stderrOldTop: -1,
            reloadIntervalId: -1,
            promises: new PromiseKeeper(),
            appType: undefined,
            webLink: undefined
        };
        this.props.setPageTitle(shortUUID(this.jobId));
    }

    get jobId(): string {
        return this.props.match.params.jobId;
    }

    public async componentDidUpdate() {
        if (!this.state.appType && this.state.app) {
            const {name, version} = this.state.app;
            const {response} = await this.state.promises.makeCancelable(
                Cloud.get<WithAppInvocation>(`/hpc/apps/${encodeURI(name)}/${encodeURI(version)}`)
            ).promise;
            this.setState(() => ({appType: response.invocation.applicationType}));
        }
    }

    public componentDidMount() {
        const reloadIntervalId = window.setTimeout(() => this.retrieveStdStreams(), 1_000);
        this.setState(() => ({reloadIntervalId}));
    }

    public componentWillUnmount() {
        if (this.state.reloadIntervalId) window.clearTimeout(this.state.reloadIntervalId);
        this.state.promises.cancelPromises();
    }

    private scrollIfNeeded() {
        if (!this.stdoutEl || !this.stderrEl) return;

        if (this.stdoutEl.scrollTop === this.state.stdoutOldTop || this.state.stderrOldTop === -1) {
            this.stdoutEl.scrollTop = this.stdoutEl.scrollHeight;
        }

        if (this.stderrEl.scrollTop === this.state.stderrOldTop || this.state.stderrOldTop === -1) {
            this.stderrEl.scrollTop = this.stderrEl.scrollHeight;
        }

        const outTop = this.stdoutEl.scrollTop;
        const errTop = this.stderrEl.scrollTop;

        this.setState(() => ({
            stdoutOldTop: outTop,
            stderrOldTop: errTop
        }));
    }

    private async retrieveStdStreams() {
        if (this.state.complete) {
            this.retrieveStateWhenCompleted();
            return;
        } else if (this.state.appState === AppState.RUNNING) {
            if (this.state.appType === "VNC") {
                this.props.setLoading(false);
                this.setState({webLink: `/novnc?jobId=${this.jobId}`})
            } else if (this.state.appType === "WEB" && !this.state.webLink) {
                this.props.setLoading(false);
                const {response} = await this.state.promises.makeCancelable(
                    Cloud.get(`/hpc/jobs/query-web/${this.jobId}`)
                ).promise;
                this.setState(() => ({webLink: response.path}));
            }
        }
        try {
            this.props.setLoading(true);
            const {response} = await this.state.promises.makeCancelable(
                Cloud.get(hpcJobQuery(this.jobId, this.state.stdoutLine, this.state.stderrLine))
            ).promise;

            this.setState(() => ({
                stdout: this.state.stdout.concat(response.stdout),
                stderr: this.state.stderr.concat(response.stderr),
                stdoutLine: response.stdoutNextLine,
                stderrLine: response.stderrNextLine,

                app: response.metadata,
                status: response.status,
                appState: response.state,
                complete: response.complete,
                outputFolder: response.outputFolder
            }));

            this.scrollIfNeeded();
            if (response.complete) this.retrieveStateWhenCompleted();
            else {
                const reloadIntervalId = window.setTimeout(() => this.retrieveStdStreams(), 1_000);
                this.setState(() => ({reloadIntervalId}));
            }
        } catch (e) {
            if (!e.isCanceled)
                snackbarStore.addSnack({
                    message: "An error occurred retrieving Information and Output from the job.",
                    type: SnackType.Failure
                });
        } finally {
            this.props.setLoading(false);
        }
    }

    private retrieveStateWhenCompleted() {
        if (!this.state.complete) return;
        window.clearTimeout(this.state.reloadIntervalId);
    }

    private renderProgressPanel = () => (
        <Panel>
            <StepGroup>
                <StepTrackerItem
                    stateToDisplay={AppState.VALIDATED}
                    currentState={this.state.appState}/>
                <StepTrackerItem
                    stateToDisplay={AppState.PREPARED}
                    currentState={this.state.appState}/>
                <StepTrackerItem
                    stateToDisplay={AppState.SCHEDULED}
                    currentState={this.state.appState}/>
                <StepTrackerItem
                    stateToDisplay={AppState.RUNNING}
                    currentState={this.state.appState}/>
                <StepTrackerItem
                    stateToDisplay={AppState.TRANSFER_SUCCESS}
                    currentState={this.state.appState}/>
            </StepGroup>
        </Panel>
    );

    private renderInfoPanel() {
        const {app} = this.state;
        if (app === undefined) return null;

        let entries = [
            {key: "Application", value: `${app.title} v${app.version}`},
            {key: "Status", value: this.state.status},
        ];

        let domEntries = entries.map(it => <Box pt="0.8em" pb="0.8em" key={it.key}><b>{it.key}</b>: {it.value}</Box>);

        switch (this.state.appState) {
            case AppState.SUCCESS:
                domEntries.push(
                    <Box key={AppState.SUCCESS} pt="0.8em" pb="0.8em">
                        Application has completed successfully.
                        Click <Link to={fileTablePage(this.state.outputFolder!)}>here</Link> to go to the output.
                    </Box>
                );
                break;
        }

        return (
            <Panel>
                <Heading.h4>Job Information</Heading.h4>
                <Card height="auto" p="14px 14px 14px 14px">
                    <List>
                        {domEntries}
                    </List>
                </Card>
            </Panel>
        );
    }

    private renderStreamPanel() {
        if (this.state.complete && this.state.stdout === "" && this.state.stderr === "") return null;
        return (
            <Box width="100%" mt={24}>
                <Heading.h4>
                    Standard Streams
                    &nbsp;
                    <Dropdown>
                        <Icon name="info" color="white" color2="black" size="1em"/>
                        <DropdownContent
                            width="400px"
                            visible
                            colorOnHover={false}
                            color="white"
                            backgroundColor="black"
                        >
                            <TextSpan fontSize={1}>
                                Streams are collected
                                from <code>stdout</code> and <code>stderr</code> of your application.
                            </TextSpan>
                        </DropdownContent>
                    </Dropdown>
                </Heading.h4>
                <Flex flexDirection="column">
                    <Box width={1} backgroundColor="midGray" mt={"12px"} pl={"12px"}>
                        <Heading.h5>Output</Heading.h5>
                    </Box>
                    <Box width={1} backgroundColor="lightGray">
                        <Stream ref={el => this.stdoutEl = el}><code>{this.state.stdout}</code></Stream>
                    </Box>
                    <Box width={1} backgroundColor="midGray" mt={"12px"} pl={"12px"}>
                        <Heading.h5>Information</Heading.h5>
                    </Box>
                    <Box width={1} backgroundColor="lightGray">
                        <Stream ref={el => this.stderrEl = el}><code>{this.state.stderr}</code></Stream>
                    </Box>
                </Flex>
            </Box>
        );
    }

    private renderFilePanel() {
        if (this.state.outputFolder === "" || this.state.appState !== AppState.SUCCESS) return null;

        return (
            <Panel>
                <Heading.h4>Output Files</Heading.h4>
                <EmbeddedFileTable path={this.state.outputFolder}/>
            </Panel>
        );
    }

    private renderWebLink() {
        const {appType} = this.state;
        if (this.state.appState === AppState.RUNNING && this.state.webLink)
            if (appType === "WEB")
                return (<ExternalLink href={this.state.webLink}>
                    <Button color="green">Go to web interface</Button>
                </ExternalLink>);
            else if (appType === "VNC")
                return <Link to={this.state.webLink}><Button color="green">Go to web interface</Button></Link>;
        return null;
    }

    private cancelJob() {
        cancelJobDialog({
            jobId: this.jobId,
            onConfirm: async () => {
                try {
                    await cancelJob(Cloud, this.jobId)
                } catch (e) {
                    snackbarStore.addFailure(errorMessageOrDefault(e, "An error occurred cancelling the job"));
                }
            }
        });
    }

    private renderCancelButton() {
        if (!inCancelableState(this.state.appState) || !this.state.app) return null;
        return <Button ml="8px" color="red" onClick={() => this.cancelJob()}>Cancel job</Button>
    }

    public render() {
        return <MainContainer
            main={this.state.app ?
                <ContainerForText>
                    {this.renderProgressPanel()}
                    {this.renderInfoPanel()}
                    {this.renderFilePanel()}
                    <Spacer left={this.renderWebLink()} right={this.renderCancelButton()}/>
                    {this.renderStreamPanel()}
                </ContainerForText> : <LoadingIcon size={24}/>}
        />
    }
}

const stateToOrder = (state: AppState): 0 | 1 | 2 | 3 | 4 | 5 => {
    switch (state) {
        case AppState.VALIDATED:
            return 0;
        case AppState.PREPARED:
            return 1;
        case AppState.SCHEDULED:
            return 2;
        case AppState.RUNNING:
            return 3;
        case AppState.TRANSFER_SUCCESS:
            return 4;
        case AppState.SUCCESS:
            return 5;
        case AppState.FAILURE:
            return 5;
        default:
            return 0;
    }
};

const isStateComplete = (state: AppState, currentState: AppState) =>
    stateToOrder(state) < stateToOrder(currentState);

const stateToTitle = (state: AppState): string => {
    switch (state) {
        case AppState.FAILURE:
            return "Failure";
        case AppState.PREPARED:
            return "Pending";
        case AppState.RUNNING:
            return "Running";
        case AppState.SCHEDULED:
            return "Scheduled";
        case AppState.SUCCESS:
            return "Success";
        case AppState.TRANSFER_SUCCESS:
            return "Transferring";
        case AppState.VALIDATED:
            return "Validated";
        default:
            return "Unknown";
    }
};

const StepTrackerItem: React.FunctionComponent<{ stateToDisplay: AppState, currentState: AppState }> = ({
                                                                                                            stateToDisplay, currentState
                                                                                                        }) => {
    const active = stateToDisplay === currentState;
    const complete = isStateComplete(stateToDisplay, currentState);
    const failed = currentState === AppState.FAILURE;
    return (
        <Step active={active}>
            {complete ?
                <Icon name={failed ? "close" : "check"} color={failed ? "red" : "green"} mr="0.7em" size="30px"/> :
                <JobStateIcon state={stateToDisplay} mr="0.7em" size="30px"/>
            }
            <TextSpan fontSize={3}>{stateToTitle(stateToDisplay)}</TextSpan>
        </Step>
    );
};

const Stream = styled.pre`
    height: 500px;
    overflow: auto;
`;

Stream.displayName = "Stream";

const mapDispatchToProps = (dispatch: Dispatch): DetailedResultOperations => ({
    setLoading: loading => dispatch(setLoading(loading)),
    setPageTitle: jobId => dispatch(updatePageTitle(`Results for Job: ${jobId}`)),
    setRefresh: refresh => dispatch(setRefreshFunction(refresh)),
});

export default connect(null, mapDispatchToProps)(DetailedResult);
