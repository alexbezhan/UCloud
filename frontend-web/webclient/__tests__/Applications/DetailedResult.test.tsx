import {configure} from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import {createMemoryHistory} from "history";
import * as React from "react";
import {Provider} from "react-redux";
import {create} from "react-test-renderer";
import {ThemeProvider} from "styled-components";
import DetailedResult from "../../app/Applications/DetailedResult";
import theme from "../../app/ui-components/theme";
import {store} from "../../app/Utilities/ReduxUtilities";
configure({adapter: new Adapter()});



describe("Detailed Result", () => {
    test("Mount DetailedResult", () => {
        expect(create(
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    <DetailedResult
                        history={createMemoryHistory()}
                        match={{
                            url: "", path: "", params: {jobId: "J0B1D"},
                            isExact: true
                        }}
                    />
                </ThemeProvider>
            </Provider>
        )).toMatchSnapshot();
    });
});
