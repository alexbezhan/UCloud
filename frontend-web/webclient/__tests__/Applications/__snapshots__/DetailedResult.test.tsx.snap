// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detailed Result Mount DetailedResult 1`] = `
.c3 {
  width: 24px;
  height: 24px;
  margin: 20px auto;
  -webkit-animation-name: jkjquP;
  animation-name: jkjquP;
}

.c3 > svg path:nth-child(1) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.04s;
  animation-delay: -0.04s;
}

.c3 > svg path:nth-child(2) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.08s;
  animation-delay: -0.08s;
}

.c3 > svg path:nth-child(3) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.12s;
  animation-delay: -0.12s;
}

.c3 > svg path:nth-child(4) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.16s;
  animation-delay: -0.16s;
}

.c3 > svg path:nth-child(5) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.2s;
  animation-delay: -0.2s;
}

.c3 > svg path:nth-child(6) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.24s;
  animation-delay: -0.24s;
}

.c3 > svg path:nth-child(7) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.28s;
  animation-delay: -0.28s;
}

.c3 > svg path:nth-child(8) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.32s;
  animation-delay: -0.32s;
}

.c3 > svg path:nth-child(9) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.36s;
  animation-delay: -0.36s;
}

.c3 > svg path:nth-child(10) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.4s;
  animation-delay: -0.4s;
}

.c3 > svg path:nth-child(11) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.44s;
  animation-delay: -0.44s;
}

.c3 > svg path:nth-child(12) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.48s;
  animation-delay: -0.48s;
}

.c3 > svg path:nth-child(13) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.52s;
  animation-delay: -0.52s;
}

.c3 > svg path:nth-child(14) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.56s;
  animation-delay: -0.56s;
}

.c3 > svg path:nth-child(15) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.6s;
  animation-delay: -0.6s;
}

.c3 > svg path:nth-child(16) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.64s;
  animation-delay: -0.64s;
}

.c3 > svg path:nth-child(17) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.68s;
  animation-delay: -0.68s;
}

.c3 > svg path:nth-child(18) {
  -webkit-animation: jkjquP 0.72s linear infinite;
  animation: jkjquP 0.72s linear infinite;
  -webkit-animation-delay: -0.72s;
  animation-delay: -0.72s;
}

.c0 {
  margin-left: 190px;
  padding-top: 48px;
  padding-bottom: 14px;
  padding-left: 14px;
  padding-right: 0;
  background-color: #fff;
}

.c1 {
  padding-top: 14px;
  padding-right: 14px;
}

@media screen and (min-width:768px) and (max-width:1023px) {
  .c2 {
    display: none;
  }
}

@media screen and (min-width:1024px) and (max-width:1279px) {
  .c2 {
    display: none;
  }
}

@media screen and (min-width:1280px) {
  .c2 {
    display: none;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    />
    <div
      className="c3"
      height={24}
      width={24}
    >
      <svg
        height={24}
        id="svg"
        version="1.1"
        viewBox="0 0 220 220"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      >
        <path
          d="M80.296,0.12l14.673,54.761l40.088,-40.088l-54.761,-14.673Z"
        />
        <path
          d="M149.73,69.554l-14.673,-54.761l-40.088,40.088l54.761,14.673Z"
        />
        <path
          d="M189.818,29.466l-54.761,-14.673l14.673,54.761l40.088,-40.088Z"
        />
        <path
          d="M204.491,84.228l-14.673,-54.762l-40.088,40.088l54.761,14.674Z"
        />
        <path
          d="M164.403,124.316l40.088,-40.088l-54.761,-14.674l14.673,54.762Z"
        />
        <path
          d="M219.165,138.989l-14.674,-54.761l-40.088,40.088l54.762,14.673Z"
        />
        <path
          d="M179.077,179.077l-14.674,-54.761l54.762,14.673l-40.088,40.088Z"
        />
        <path
          d="M124.316,164.403l54.761,14.674l-14.674,-54.761l-40.087,40.087Z"
        />
        <path
          d="M138.989,219.165l40.088,-40.088l-54.761,-14.674l14.673,54.762Z"
        />
        <path
          d="M84.228,204.491l54.761,14.674l-14.673,-54.762l-40.088,40.088Z"
        />
        <path
          d="M69.554,149.73l14.674,54.761l40.088,-40.088l-54.762,-14.673Z"
        />
        <path
          d="M29.466,189.818l54.762,14.673l-14.674,-54.761l-40.088,40.088Z"
        />
        <path
          d="M14.793,135.057l14.673,54.761l40.088,-40.088l-54.761,-14.673Z"
        />
        <path
          d="M54.881,94.969l-40.088,40.088l54.761,14.673l-14.673,-54.761Z"
        />
        <path
          d="M0.12,80.296l14.673,54.761l40.088,-40.088l-54.761,-14.673Z"
        />
        <path
          d="M40.208,40.208l-40.088,40.088l54.761,14.673l-14.673,-54.761Z"
        />
        <path
          d="M94.969,54.881l-54.761,-14.673l14.673,54.761l40.088,-40.088Z"
        />
        <path
          d="M80.296,0.12l-40.088,40.088l54.761,14.673l-14.673,-54.761Z"
        />
      </svg>
    </div>
  </div>
</div>
`;
